### DISCUSSION TITLE
How to use updateOne of createEntityAdapter()?

### DISCUSSION THREAD

**Discussion by colorcube** [state-adapt/state-adapt#38] [Q&A]
I can't figure out what's the right way to use updateOne created by createEntityAdapter()?

I have a list of entities. There are actions on single entities (set flags, ...).
The action should trigger a http request and finally the entity should be updated in the store.

I've seen examples which return getAction() instead of toSource
{id, changes} should be used as payload, but there are examples with [id, changes]

A simple example (excerpt) would be nice

**Comment by mfp22** [comment]
Go to https://state-adapt.github.io/adapters/core#createEntityAdapter
And change the state change to `updateOne`
And you can see what the payload needs to look like.

It is `[id changes]`. That's different from NgRx, but I think `{ id, changes}` is unnecessarily verbose, especially when calling it from a template. When reading the code, it's easy to understand what's happening. It's usually like `updateOne([id, { name: $event }])`.

Also it makes it more consistent with the other `one` state changes. They take in a payload for a single entity state change, and it doesn't make sense to name it anything other than payload. So it was going to be different from ngrx entity either way. Might as well go with the slimmer syntax.

If you're ever uncertain about the types and plugging in a source isn't giving a clean TypeScript error, try talking to the store directly, like `this.store.updateOne(`. That should tell you what it expects.

**Comment by colorcube** [comment]
I think my problem has to do with using a different ID than 'id'.
With my code updateOne() get's not created and is undefined and I have some type problems.

I used Books (2023-fidelity-workshop) as example:

This adds some functionality to use updateBooksOne and it works
https://github.com/colorcube/2023-fidelity-workshop/commit/7e733211d53156294cb9d9434973444e61367af0

Here I changed the id to bookId
https://github.com/colorcube/2023-fidelity-workshop/commit/7e733211d53156294cb9d9434973444e61367af0

It's broken.

I guess EntityState, createEntityState and createEntityAdapter needs a type for the id. But my attempts give not clean results.

**Comment by colorcube** [comment]
Observation ...

```typescript
const bookAdapterWithoutUpdate = createAdapter<BookModel>()({})();

const bookAdapterWithUpdate = joinAdapters<BookModel, keyof BookModel>()({})();
```

bookAdapterWithUpdate has an update reaction.

When bookAdapterWithUpdate is passed to createLabelEntityAdapter() as adapter the update* reactions are generated.

This was not obvious to me.

This types problem in the Books example above persists.

**Comment by colorcube** [comment]
I made a mistake with the id selector:

```typescript
const bookAdapter = joinAdapters<BookModel, keyof BookModel>()({
  selectors: {
    id: (s) => s.bookId,
  }
})(); 
```

This is right:

```typescript
const bookAdapter = joinAdapters<BookModel, keyof BookModel>()({
    // Selectors
    id: (s) => s.bookId,
})(); 
```

But that doesn't solve the type problem.

For my code I hacked create-entity-adapter.function.ts to make the typing and id work.
That showed an additional problem with the id selector.

bookAdapter provides an id selector, but that is a memoized selector, which means always the same id is returned for createEntityAdapter.

This hack works
```typescript
    // entitySelectors.id = entitySelectors.id || ((entity: Entity) => (entity as any).id as Id);
    entitySelectors.id = ((entity: Entity) => (entity as any).bookId as Id);
```

**Reply by mfp22** [reply]
You probably know this, but for anyone else who might see this, it should be 

```typescript
const bookAdapter = joinAdapters<BookModel, keyof BookModel>()({
    // adapters
})({
    // Selectors
    id: (s) => s.bookId,
})();
```

For `createAdapter` I decided not to have `update` as a default reaction, because I would have to figure out how to make sure the state is shaped like an ordinary object and not something else. I could just include it by default, but then someone might try to `update` an array and not realize what's wrong. For object states, I think people should probably use `joinAdapters`. Otherwise it isn't too much work to define your own `update: (state, changes: Partial<State>) => ({...state, ...changes}),`

I'll clone your repo later today and investigate the type problem. Thanks for being patient with this. The entity adapter is the least tested part of StateAdapt, and I personally don't usually use the `update` function. I think it sort of puts knowledge of the state shape into the consumers, and I like the sources to be "pure", just representing stuff that happened... not always possible. Anyway, `update` is convenient sometimes.

**Reply by colorcube** [reply]
Right, my mistake with the selector. But fixing it results in following error:

`Property 'bookId' does not exist on type 'SelectorReturnTypes<BookModel, {}>'`

Thanks for looking into it

**Reply by mfp22** [reply]
You need an adapter for the `bookId`

```typescript
type ExcludedBookModelProps = 'name' | 'earnings' | 'description';
const bookAdapter = joinAdapters<BookModel, ExcludedBookModelProps>()({
  // adapters
  bookId: createAdapter<string>()({}),
})({
  // Selectors
  id: (s) => s.bookId,
})(); // https://state-adapt.github.io/docs/core#joinadapters
```

`joinAdapters` can't create selectors based on types alone. The JavaScript needs an actual property name to use to generate state change functions and selectors. Now the `id` selector isn't giving that working. 

There's still an issue. I'm still looking

**Reply by mfp22** [reply]
Next issue: Need to pass in the alternative id `EntityState` and `createEntityAdapter`

```typescript
export interface State {
  activeBookId: string | null;
  isLoading: boolean;
  error: object | null;
  books: EntityState<BookModel, 'bookId'>;
}

// ... 

const booksAdapter = createEntityAdapter<BookModel, 'bookId'>()(bookAdapter);
```

**Reply by mfp22** [reply]
This one _seems_ to be an issue with the `createEntityAdapter` types: 

```error
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Record<"bookId", BookModel>'.
```

**Reply by mfp22** [reply]
Yes, in `node_modules/@state-adapt/core/adapters/lib/create-entity-adapter.function.d.ts`

Line 8 needs to be `entities: Record<Entity[Id], Entity>;`
Line 45 needs to be `entities: (state: State) => Record<Entity[Id], Entity>;` 

The app still isn't compiling, so I will see what else is going on

**Reply by mfp22** [reply]
Okay, this is a problem... Pretty much everywhere I am using the alternate `Id` in the entity types I should be using `Entity[Id]`. This is the result of my mistake:
<img width="621" alt="image" src="https://github.com/state-adapt/state-adapt/assets/13700926/64f13a8e-228b-46c7-9621-ad116e6134ef">

Pretty silly in hindsight.

**Reply by mfp22** [reply]
Well, I made a few changes, so I might as well give you the entire new file for `create-entity-adapter.function.d.ts`:

```typescript
import { Selectors, ReactionsWithSelectors, Adapter, BasicAdapterMethods, Flat } from '@state-adapt/core';
import { PrefixedAfterVerb } from '@state-adapt/core';
declare type Index = number | string | symbol;
export interface EntityState<Entity, Id extends Index = Entity extends {
    id: Index;
} ? Entity['id'] : never> {
    ids: Id[];
    entities: Record<Entity[Id], Entity>;
}
export declare type EntityAdapterOptions<A extends Adapter<any, any, any>> = {
    adapter: A;
};
export declare type BooleanSelectorKeys<S extends Selectors<any>> = keyof {
    [K in Extract<keyof S, string>]: S[K] extends (state: any) => boolean ? K : never;
};
export declare function createEntityState<Entity, Id extends Index = Entity extends {
    id: Index;
} ? Entity['id'] : never, State extends EntityState<Entity, Id> = EntityState<Entity, Id>>(state?: Partial<State>): EntityState<Entity, Id>;
declare type EntityStateReactions<Entity, Id extends Index, State extends EntityState<Entity, Id>, R extends ReactionsWithSelectors<Entity, any>, Filters extends string[]> = Flat<{
    addOne: (state: State, payload: Entity) => State;
    addMany: (state: State, payload: Entity[]) => State;
    setOne: (state: State, payload: Entity) => State;
    setMany: (state: State, payload: Entity[]) => State;
    setAll: (state: State, payload: Entity[]) => State;
    removeOne: (state: State, payload: Entity[Id]) => State;
    removeMany: (state: State, payload: Entity[Id][]) => State;
    removeAll: (state: State) => State;
    updateOne: (state: State, payload: [Entity[Id], Partial<Entity>]) => State;
    upsertOne: (state: State, payload: Entity) => State;
    upsertMany: (state: State, payload: Entity[]) => State;
} & BasicAdapterMethods<State> & {
    [K in Extract<keyof R, string> as K extends typeof ingoredKeys[number] ? never : PrefixedAfterVerb<Exclude<K, 'set' | 'reset'>, Filters[number] | 'all'>]: R[K] extends (state: any, payload: any, initialState: any) => any ? (state: State, payload: Parameters<R[K]>[1] extends void ? void : Parameters<R[K]>[1], initialState: State) => State : never;
} & {
    [K in Extract<keyof R, string> as K extends typeof ingoredKeys[number] ? never : PrefixedAfterVerb<K, 'one'>]: R[K] extends (state: any, payload: any, initialState: any) => any ? (state: State, payload: Parameters<R[K]>[1] extends void ? Id : [Entity[Id], Parameters<R[K]>[1]], initialState: State) => State : never;
} & {
    [K in Extract<keyof R, string> as K extends typeof ingoredKeys[number] ? never : PrefixedAfterVerb<K, 'many'>]: R[K] extends (state: any, payload: any, initialState: any) => any ? (state: State, payload: Parameters<R[K]>[1] extends void ? Id[] : [Entity[Id], Parameters<R[K]>[1]][], initialState: State) => State : never;
}>;
declare type EntityStateSelectors<Entity, Id extends Index, State extends EntityState<Entity, Id>, Filters extends string[], Sorters extends string[]> = {
    all: (state: State) => Entity[];
    count: (state: State) => number;
    ids: (state: State) => Entity[Id][];
    entities: (state: State) => Record<Entity[Id], Entity>;
} & {
    [K in Filters[number]]: (state: any) => Entity[];
} & {
    [K in Filters[number] as `${K}Count`]: (state: any) => number;
} & {
    [K in Filters[number] as `allAre${Capitalize<K>}`]: (state: any) => boolean;
} & {
    [K in `${Filters[number] | 'all'}By${Capitalize<Sorters[number]>}`]: (state: any) => Entity[];
};
declare const ingoredKeys: readonly ["selectors", "set"];
export declare function createEntityAdapter<Entity, Id extends Index = Entity extends {
    id: Index;
} ? Entity['id'] : never, State extends EntityState<Entity, Id> = EntityState<Entity, Id>>(): <S extends Selectors<Entity>, R extends ReactionsWithSelectors<Entity, S>, Filters extends Extract<keyof S, string>[] = never[], Sorters extends Extract<keyof S, string>[] = never[]>(adapter: Adapter<Entity, S, R>, options?: {
    filters?: Filters | undefined;
    sorters?: Sorters | undefined;
    useCache?: boolean | undefined;
} | undefined) => EntityStateReactions<Entity, Id, State, R, Filters> & {
    selectors: EntityStateSelectors<Entity, Id, State, Filters, Sorters>;
};
export {};

```

**Reply by mfp22** [reply]
The app builds and runs now, but there are some issues with deleting and editing books. Maybe you had some work in progress? Let me know if you can pinpoint where the remaining issues could be.

**Reply by colorcube** [reply]
Editing worked definitely in this commit
https://github.com/colorcube/2023-fidelity-workshop/commit/7e733211d53156294cb9d9434973444e61367af0

I didn't test deleting

Click on the check marks sets the description to 'SPECIAL'.

That's it. No more changes are in progress.

**Reply by mfp22** [reply]
It might be a server problem

**Reply by colorcube** [reply]
Delete worked too
Not sorting, but that didn't work before.

**Reply by mfp22** [reply]
Probably not though. I'm looking more now

**Reply by mfp22** [reply]
The issue is that the fetchAll() books request is returning books with `{ id: ` instead of `bookId`

**Reply by mfp22** [reply]
Anyway, I created this issue https://github.com/state-adapt/state-adapt/issues/41 for the types issue. Anyone finding this can use the file code I provided above until StateAdapt 1.1.1 is released with the bugfix.

**Reply by colorcube** [reply]
Right, I forgot that

**Reply by colorcube** [reply]
If you like you can push the changes in my repo.

Could you post the changed create-entity-adapter.function.ts file please?

**Reply by mfp22** [reply]
I posted it above. I'm having trouble pushing to that repo

**Reply by colorcube** [reply]
You posted the .d.ts file

**Reply by mfp22** [reply]
Here's the change I made in the service though 

```typescript
interface ApiBookModel {
  id: string;
  name: string;
  earnings: string;
  description?: string;
}

@Injectable({
  providedIn: 'root',
})
export class BooksService {
  constructor(private http: HttpClient) {}

  fetchAll() {
    return this.http.get<ApiBookModel[]>(BASE_URL).pipe(
      map((books) =>
        books.map((book) => ({
          bookId: book.id,
          name: book.name,
          earnings: book.earnings,
          description: book.description,
        }))
      )
    );
  }
```

So now I've posted all my changes here

**Reply by colorcube** [reply]
https://github.com/colorcube/2023-fidelity-workshop/invitations

**Reply by mfp22** [reply]
https://github.com/colorcube/2023-fidelity-workshop/commit/6eab312d23eaa653a1b0c815f50b7b66454fcadb

**Reply by colorcube** [reply]
There's still a small bug.
When editing an entry and save it, a duplicate entry with the same bookId is generated but with an additional id.
But that comes from the backend and can be ignored.

**Comment by colorcube** [comment]
It's a bit tricky to override .d.ts

This worked for me

tsconfig.json (excerpt):

```json
{
    "compilerOptions": {
        "paths": {
            "@state-adapt/core/adapters": ["typings/@state-adapt/core/adapters"],
        }
    }
}
```

Copy d.ts files from node_modules/@state-adapt/core/adapters to (a newly created) typings folder:

```
typings
└── @state-adapt
    └── core
        └── adapters
            ├── index.d.ts
            └── lib
                ├── boolean.adapter.d.ts
                ├── create-entity-adapter.function.d.ts
                ├── number.adapter.d.ts
                └── string.adapter.d.ts
```

Put the type definition from [above](https://github.com/state-adapt/state-adapt/discussions/38#discussioncomment-6442295) into create-entity-adapter.function.d.ts (is nearly the same)

Overriding  create-entity-adapter.function.d.ts only won't work

**Reply by mfp22** [reply]
Really? Maybe it's a web storm thing. Did you mention you were using webstorm? In VS Code I can edit the d.ts file within node_modules and it just works.

**Reply by colorcube** [reply]
Patching node_modules works of course. 
But my stuff will be build with CI and npm i

**Reply by mfp22** [reply]
Oh right, so you wanted something a little more permanent. Well, it should only be a couple more days before I get 1.1.1 out with this fix.

**Comment by mfp22** [comment]
The TypeScript issue should be fixed as of 1.2.1. 

This was the commit: 29ddbca

It was weirder than I thought it would be, but the improved tests went from failing to passing, and the entity adapter on the docs site was working. https://state-adapt.github.io/adapters/core#createEntityAdapter
